crates/red_knot_python_semantic/src/semantic_index/use_def/symbol_state.rs

```diff
@@ -0,0 +1,362 @@
+const INLINE_DEFINITION_BLOCKS: usize = 1;
+
+/// A [`BitSet`] of [`ScopedDefinitionId`], representing visible definitions of a symbol in a scope.
+type Definitions = BitSet<INLINE_DEFINITION_BLOCKS>;
+type DefinitionsIterator<'a> = BitSetIterator<'a, INLINE_DEFINITION_BLOCKS>;
+
+/// Can reference this * 64 total constraints inline; more will fall back to the heap.
+const INLINE_CONSTRAINT_BLOCKS: usize = 1;
+
+/// Can handle this many visible definitions per symbol at a given time inline.
+const INLINE_VISIBLE_DEFINITIONS_PER_SYMBOL: usize = 4;
+
+/// A [`BitSetArray`];one [`BitSet`] of applicable [`ScopedConstraintId`] per visible definition.
+type Constraints = BitSetArray<INLINE_CONSTRAINT_BLOCKS, INLINE_VISIBLE_DEFINITIONS_PER_SYMBOL>;
+type ConstraintsIterator<'a> =
+    BitSetArrayIterator<'a, INLINE_CONSTRAINT_BLOCKS, INLINE_VISIBLE_DEFINITIONS_PER_SYMBOL>;
+
+/// Visible definitions and narrowing constraints for a single symbol at some point in control flow.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(super) struct SymbolState {
+    /// [`BitSet`]: which [`ScopedDefinitionId`] are visible for this symbol?
+    visible_definitions: Definitions,
+
+    /// For each definition, which [`ScopedConstraintId`] apply?
+    ///
+    /// This is a [`BitSetArray`] which should always have one [`BitSet`] of constraints per
+    /// definition in `visible_definitions`.
+    constraints: Constraints,
+
+    /// Could the symbol be unbound at this point?
+    may_be_unbound: bool,
+}
+
+/// A single [`ScopedDefinitionId`] with an iterator of its applicable [`ScopedConstraintId`].
+#[derive(Debug)]
+pub(super) struct DefinitionIdWithConstraints<'a> {
+    pub(super) definition: ScopedDefinitionId,
+    pub(super) constraint_ids: ConstraintIdIterator<'a>,
+}
+
+impl SymbolState {
+    /// Return a new [`SymbolState`] representing an unbound symbol.
+    pub(super) fn unbound() -> Self {
+        Self {
+            visible_definitions: Definitions::default(),
+            constraints: Constraints::default(),
+            may_be_unbound: true,
+        }
+    }
+
+    /// Return a new [`SymbolState`] representing a symbol with a single visible definition.
+    pub(super) fn with(definition_id: ScopedDefinitionId) -> Self {
+        Self {
+            visible_definitions: Definitions::with(definition_id.into()),
+            constraints: Constraints::of_size(1),
+            may_be_unbound: false,
+        }
+    }
+
+    /// Add Unbound as a possibility for this symbol.
+    pub(super) fn add_unbound(&mut self) {
+        self.may_be_unbound = true;
+    }
+
+    /// Add given constraint to all currently-visible definitions.
+    pub(super) fn add_constraint(&mut self, constraint_id: ScopedConstraintId) {
+        self.constraints.insert_in_each(constraint_id.into());
+    }
+
+    /// Merge two [`SymbolState`] and return the result.
+    pub(super) fn merge(a: &SymbolState, b: &SymbolState) -> SymbolState {
```

Comment:
I'm unsure if we can make use of this but:

We should change `builder.flow_merge` to take an owned snapshot. I didn't had to insert a single `snapshot.clone`. This could help us to possibly avoid some `clone` calls because it seems we can consume the value.

Reply:
I think I used to have this but I changed it to take a reference because clippy doesn't like it if it's passed by value but not consumed. And it doesn't seem like we can gain anything by consuming it; it isn't cloned anywhere in this method. We already reuse `self` in-place and merge the data from `snapshot` into it; we can't reuse both of them.

Reply:
I do think we could make use of it but it would probably complicate the logic a bit because you need two versions of `push` and a `into_iter` version for `constraints` 

For example, couldn't we avoid cloning the constraints and instead take them right from the symbol state?

Reply:
Ok, yeah, I was too tired yesterday after going through all the other comments to push this all the way through :) You're right, I can eliminate the `constraints.clone()` this way; I don't even need two variants of `push` because I can use `into_iter` for iterating both incoming constraint sets.

Sadly this doesn't seem to move the needle at all on the benchmarks, but it makes sense to do it this way so I'll leave it.

Going to go ahead and merge this, but feel free to take a look at the updated `SymbolState::merge` and `UseDefMapBuilder::merge` and tell me if you see anywhere else I'm leaving potential perf on the floor. Thanks!